sklearn/preprocessing/_encoders.py

```diff
@@ -397,6 +496,190 @@ def _compute_drop_idx(self):
+            if infreq_idx is None:
+                default_to_infrequent_mappings.append(None)
+                continue
+
+            n_cats = len(cats)
+            # infrequent indices exist
+            mapping = np.empty(n_cats, dtype=int)
+            n_infrequent_cats = infreq_idx.size
+
+            # infrequent categories are mapped to the last element.
+            n_frequent_cats = n_cats - n_infrequent_cats
+            mapping[infreq_idx] = n_frequent_cats
+
+            frequent_indices = np.setdiff1d(np.arange(n_cats), infreq_idx)
+            mapping[frequent_indices] = np.arange(n_frequent_cats)
+
+            default_to_infrequent_mappings.append(mapping)
+
+        self._default_to_infrequent_mappings = default_to_infrequent_mappings
+
+    def _map_to_infrequent_categories(self, X_int, X_mask):
+        """Map categories to infrequent categories. This modifies X_int
+        in-place.
+
+        Parameters
+        ----------
+        X_int: ndarray of shape (n_samples, n_features)
+            Integer encoded categories.
+        """
+        if not self._infrequent_enabled:
+            return
+
+        n_features = X_int.shape[1]
+        for col_idx in range(n_features):
+            infrequent_idx = self.infrequent_indices_[col_idx]
+            if infrequent_idx is None:
+                continue
+
+            X_int[~X_mask[:, col_idx], col_idx] = infrequent_idx[0]
+            if self.handle_unknown == 'auto':
+                # unknown values will be mapped to infrequent in the next for
+                # loop
+                X_mask[:, col_idx] = True
+
+        for i, mapping in enumerate(self._default_to_infrequent_mappings):
+            if mapping is None:
+                continue
+            X_int[:, i] = np.take(mapping, X_int[:, i])
+
+    def _compute_transformed_categories(self, i):
+        """Compute the transformed categories used for column `i`.
+
+        1. Dropped columns are removed.
+        2. If there are infrequent categories, the category is named
+        'infrequent'. If 'infrequent' is already a category, then then new
+        category is called 'infrequent_sklearn'.
+        """
+        cats = self.categories_[i]
+
+        if self.drop_idx_ is not None:
+            if self.drop_idx_[i] is None:
+                return cats
+            return np.delete(cats, self.drop_idx_[i])
+
+        # drop is None
+        if not self._infrequent_enabled:
+            return cats
+
+        # infrequent is enabled
+        infreq_idx = self.infrequent_indices_[i]
+        if infreq_idx is None:
+            return cats
+
+        frequent_indices = np.setdiff1d(np.arange(len(cats)), infreq_idx)
+
+        if cats.dtype.kind in 'US' and 'infrequent' in cats:
+            infrequent_cat = 'infrequent_sklearn'
+        else:
+            infrequent_cat = 'infrequent'
+        return np.r_[cats[frequent_indices],
+                     np.array([infrequent_cat], dtype=object)]
```

Comment:
I think usng `._default_to_infrequent_mappings` is a little neater (does it work?):
```suggestion
        infreq_map = self._default_to_infrequent_mappings[i]
        if infreq_map is None:
            return cats

        frequent_mask = infreq_map < infreq_map.max()

        if cats.dtype.kind in 'US' and 'infrequent' in cats:
            infrequent_cat = 'infrequent_sklearn'
        else:
            infrequent_cat = 'infrequent'
        return np.concatenate([cats[frquent_mask],
                               np.array([infrequent_cat], dtype=object)])
```